src/rpc/services/SpectateService.ts

```diff
@@ -0,0 +1,531 @@
+        if (!latestForkGenesisSnapshot) {
+            throw new Error(`No genesis snapshot found for fork ${forkId}`);
+        }
+
+        const latestBlockHeight =
+            stateManager.storage.blocks.getNextBlockHeight(forkId) - 1;
+
+        // Get the latest state proof data (only if there are blocks for same-fork update)
+        let latestStateProof: StateProofStruct | undefined;
+        let encodedState: Hash | undefined;
+
+        if (latestBlockHeight >= 0) {
+            // There are blocks, so we can do a same-fork update
+            latestStateProof = await agreementManager.getStateProof(
+                forkId,
+                latestBlockHeight
+            );
+
+            // Get the encoded state that the stateProof proves
+            if (latestStateProof.milestones.length > 0) {
+                const latestMilestone =
+                    latestStateProof.milestones[
+                        latestStateProof.milestones.length - 1
+                    ];
+                const latestSnapshot =
+                    agreementManager.getSnapshot(latestMilestone);
+                const stateHash =
+                    latestSnapshot.snapshotData.stateMachineStateHash;
+                encodedState =
+                    stateManager.storage.stateMachineStates.getStateMachineState(
+                        stateHash
+                    );
+                if (!encodedState) {
+                    throw new Error(
+                        `No encoded state found for state hash ${stateHash}`
+                    );
+                }
+            }
+        }
+
+        // Get current on-chain snapshot to start the fork traversal
+        const currentOnChainSnapshot = StateSnapshot.from(
+            await stateManager.stateChannelManagerContract.getStateSnapshot(
+                channelId
+            )
+        );
+
+        // Traverse from on-chain snapshot to latest fork, collecting disputes per dispute window
+        const disputeWindows: DisputeStruct[][] = [];
+        let currentForkId = currentOnChainSnapshot.forkId;
+        let isDisputed =
+            await stateManager.stateChannelManagerContract.isForkDisputed(
+                channelId,
+                currentForkId
+            );
+
+        while (isDisputed) {
+            // Check if reduced result already exists on-chain
+            const existingReducedResult =
+                await stateManager.stateChannelManagerContract.getReducedResult(
+                    channelId,
+                    currentForkId
+                );
+
+            if (existingReducedResult[0]) {
+                // Traverse to the reduced fork
+                currentForkId = existingReducedResult[0];
+                isDisputed =
+                    await stateManager.stateChannelManagerContract.isForkDisputed(
+                        channelId,
+                        currentForkId
+                    );
+                continue;
+            }
```

Comment:
On a first though this seems like the optimal way to do it "If it's reduced I don't need to provide the data to them to reduce it, they can just check on-chain and if it's not reduced I'll provide the data, so they can verify for themselves - keep in mind when you do it like that there could be race condition on-chain where somone reduced a disputeWindow on-chain `after you packaged the payload` and `before the spectator verifies it`, so you have to account for it in the verification code

Reply:
I'd keep it simple and not optimized (to not think about these race conditions and edge cases for now) and just provide all disputeWindows and they will run reduce on everything regardless to derive the final forId - if there's inconsistency they would just abort